File: Sources/Engine/Core/Math/Vec/Vec3.cpp
#include "Chicane/Core/Math/Vec/Vec3.reflected.hpp"

#include "Chicane/Core/Math/Vec/Vec4.hpp"

namespace Chicane
{
    constexpr Vec3::Vec3(const Vec4& inValue)
        : x(inValue.x),
          y(inValue.y),
          z(inValue.z)
    {}

    String Vec3::toString() const
    {
        return String::sprint("[%.2f, %.2f, %.2f]", x, y, z);
    }

    Vec3 Vec3::min(const Vec3& inValue) const
    {
        return glm::min(static_cast<glm::vec3>(*this), static_cast<glm::vec3>(inValue));
    }

    Vec3 Vec3::max(const Vec3& inValue) const
    {
        return glm::max(static_cast<glm::vec3>(*this), static_cast<glm::vec3>(inValue));
    }

    Vec3 Vec3::cross(const Vec3& inValue) const
    {
        return glm::cross(static_cast<glm::vec3>(*this), static_cast<glm::vec3>(inValue));
    }

    float Vec3::dot(const Vec3& inValue) const
    {
        return glm::dot(static_cast<glm::vec3>(*this), static_cast<glm::vec3>(inValue));
    }

    Vec3 Vec3::normalize() const
    {
        return glm::normalize(static_cast<glm::vec3>(*this));
    }
}

File: Includes/Chicane/Core/Script/Types/Vec3.lua
---@meta

---@class Vec3
---@field x number
---@field y number
---@field z number
Vec3 = {}

---@param x number
---@param y number
---@param z number
---@return Vec3
function Vec3(x, y, z) end

---@return Vec3
function Vec3.Zero() end

---@return Vec3
function Vec3.One() end

---@return Vec3
function Vec3.Right() end

---@return Vec3
function Vec3.Forward() end

---@return Vec3
function Vec3.Up() end

---@return string
function Vec3:toString() end

---@param value Vec3
---@return Vec3
function Vec3:min(value) end

---@param value Vec3
---@return Vec3
function Vec3:max(value) end

---@param value Vec3
---@return Vec3
function Vec3:cross(value) end

---@param value Vec3
---@return number
function Vec3:dot(value) end

---@return Vec3
function Vec3:normalize() end
